pkg/util/iptables/iptables.go

```diff
@@ -575,3 +576,252 @@ func IsNotFoundError(err error) bool {
+	Jump    Chain
+	Comment string
+	Options map[string]RuleOption
+}
+
+// Split a line into space-separated arguments, handling simple quoting
+func splitLine(line string) ([]string, error) {
+	items := strings.Split(line, " ")
+	args := make([]string, 0, len(items))
+	for i := 0; i < len(items); i++ {
+		if items[i] == "" {
+			continue
+		}
+
+		if !strings.HasPrefix(items[i], "\"") {
+			args = append(args, items[i])
+			continue
+		}
+
+		// combine until we find a closing quote
+		var arg string
+		items[i] = items[i][1:]
+		for j := i; j < len(items); j++ {
+			if len(arg) > 0 {
+				arg += " "
+			}
+			arg += items[j]
+			if strings.HasSuffix(arg, "\"") {
+				args = append(args, arg[:len(arg)-1])
+				break
+			}
+			i++
+		}
+		if i == len(items) {
+			return nil, fmt.Errorf("invalid iptables rule (unterminated quote): %v", line)
+		}
+	}
+	return args, nil
+}
+
+func addRuleOption(rule *Rule, option string, ro RuleOption) error {
+	if rule.Options == nil {
+		rule.Options = make(map[string]RuleOption)
+	}
+	if _, ok := rule.Options[option]; ok {
+		return fmt.Errorf("invalid iptables rule (multiple %q args)", option)
+	}
+	rule.Options[option] = ro
+	return nil
+}
+
+var optionAttrs = map[string]struct {
+	argRequired bool
+	canNegate   bool
+}{
+	// Argument required, cannot be negated
+	"-A":               {true, false},
+	"-j":               {true, false},
+	"-m":               {true, false},
+	"--comment":        {true, false},
+	"--name":           {true, false},
+	"--seconds":        {true, false},
+	"--dst-type":       {true, false},
+	"--set-xmark":      {true, false},
+	"--to-destination": {true, false},
+	"--mark":           {true, false},
+	"--mask":           {true, false},
+	"--mode":           {true, false},
+	"--probability":    {true, false},
+	// Argument required, can be negated
+	"-p":         {true, true},
+	"-d":         {true, true},
+	"-s":         {true, true},
+	"-i":         {true, true},
+	"-o":         {true, true},
+	"--dport":    {true, true},
+	"--src-type": {true, true},
+	// No argument required, cannot be negated
+	"--reap":    {false, false},
+	"--rsource": {false, false},
+	"--set":     {false, false},
+	// No argument required, can be negated
+	"--physdev-is-in": {false, true},
+	"--rcheck":        {false, true},
+}
+
+func parseRule(line string, allowChains []Chain, allowChainPrefixes []string) (*Rule, Chain, error) {
+	var rule Rule
+	var chain Chain
+
+	items, err := splitLine(line)
+	if err != nil {
+		return nil, "", err
+	}
+
+	for i := 0; i < len(items); i++ {
+		// negation only valid at beginning of options
+		var negated bool
+		if items[i] == "!" {
+			negated = true
+			i++
+			if i == len(items) {
+				return nil, "", fmt.Errorf("invalid iptables rule (not enough arguments): %v", line)
+			}
+		}
+
+		opt := items[i]
+		attr, ok := optionAttrs[opt]
+		if !ok {
+			return nil, "", fmt.Errorf("Unknown iptables rule option %q", opt)
+		}
+		if !attr.canNegate && negated {
```

Comment:
Do we actually care about this?  Why not just assume that a `!` always applies to the next arg.  If if doesn't work as an actual iptables command, who cares.  That simplifies this and more importantly the table of commands can become a map of `flagName : hasArg`